Fix block-connection performance regression

TheBlueMatt opened the pull request:
@JeremyRubin found a performance regression introduced by b3b3c2a5623d5c942d2b3565cc2d833c65105555 which causes us to make copies of every transaction we connect to tip a while back, but no one had gotten around to fixing it yet. By switching to using shared_ptrs for blocks in several places we not only fix the performance regression but open up the possibility of switching GetMainSignals() callbacks to background threads much more easily.

TheBlueMatt committed:
Keep blocks as shared_ptrs, instead of copying txn in ConnectTip

TheBlueMatt committed:
Make the optional pblock in ActivateBestChain a shared_ptr

sipa commented on src/main.cpp:
```diff
@@ -2799,20 +2799,40 @@ static int64_t nTimeFlush = 0;
 
+struct CompareBlockIndexByWork {
+    bool operator()(const CBlockIndex* a, const CBlockIndex* b) const {
+        return a->nChainWork < b->nChainWork;
```
Don't you need a fallback for when the chainwork is identical?

TheBlueMatt replied:
Since its only used in each ActivateBestChainStep call individually, there is only ever one thing per given height/total work, but the map came out of an older version of the code and is now useless...so I'm replacing it with a vector.

JeremyRubin commented on src/main.cpp:
```diff
@@ -2799,20 +2799,40 @@ static int64_t nTimeFlush = 0;
+// Note that since mapBlocks is sorted lowest-work to highest-work, iterating
+// over it will give you the order in which the blocks were connected
+struct BlockConnectTrace {
+    std::map<CBlockIndex*, std::shared_ptr<const CBlock>, CompareBlockIndexByWork> mapBlocks;
```
I think it may be worthwhile for you to make this a class and keep the implementation details hidden...

TheBlueMatt replied:
I added txConflicted here to make diffs of individual commits easier, which now requires a lot of access to the contents directly :/ Not sure its worth it.

JeremyRubin reviewed (commented):
utACK

couple things that could be cleaned up, but I think this is the right approach for this and is OK as is, certainly better than txChanged.

TheBlueMatt commented on src/main.cpp:
```diff
@@ -5829,7 +5855,8 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
         vRecv >> resp;
 
-        CBlock block;
+        CBlock *pblock = new CBlock();
```
The shared_ptrs are to a const CBlock, and IIRC you cant blindly cast a shared_ptr<CBlock> to a shared_ptr<const CBlock>, but you need to access the block as non-const, and I kinda prefer this over a const_cast.

JeremyRubin replied:
implicit copy-cast should work...

```
#include <memory>
int main() {
    auto a = std::make_shared<int>(10);
    std::shared_ptr<const int> b = a;
}
```

TheBlueMatt committed:
Add struct to track block-connect-time-generated info for callbacks

TheBlueMatt committed:
Create a shared_ptr for the block we're connecting in ActivateBCS

TheBlueMatt commented:
OK, ended up rewriting the PR quite a bit, cleaning it up and resolving the nits folks had raised - it now no longer has this crazy map thing that can be used as both a paramter and a return value.

JeremyRubin commented:
re-utack, much cleaner!

sipa commented on src/main.cpp:
```diff
@@ -2974,6 +2975,8 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
                     state = CValidationState();
                     fInvalidFound = true;
                     fContinue = false;
+                    // If we didn't actually connect the block, don't notify listeners about it
```
This logic needs to be duplicated in the disconnect logic above. It's possible that an invalid block is received which triggers a reorg. So we start from chain A-B, and receive A-C-D, with D invalid. We'll first disconnect B, then successfully connect C, but fail when connecting D. We'll loop back then, disconnect C, and reconnect B. Your code will (I believe) report both C and B as newly connected, while it should be nothing at all.

Perhaps it's easier to do a filtering step in ActivateBestChain that removes the errorneously blocks in the trace.

sipa replied:
Ok (answered elsewhere).

sipa commented on src/main.cpp:
```diff
@@ -2806,11 +2806,13 @@ struct ConnectTrace {
     std::vector<std::pair<CBlockIndex*, std::shared_ptr<const CBlock> > > blocksConnected;
 };
 
+
```
Unintentional new lines?

sipa commented on src/main.cpp:
```diff
@@ -3740,7 +3766,7 @@ static bool AcceptBlock(const CBlock& block, CValidationState& state, const CCha
 }
 
-bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, CNode* pfrom, const CBlock* pblock, bool fForceProcessing, const CDiskBlockPos* dbp)
+bool ProcessNewBlock(CValidationState& state, const CChainParams& chainparams, CNode* pfrom, std::shared_ptr<const CBlock>& pblock, bool fForceProcessing, const CDiskBlockPos* dbp)
```
In my opinion, passing by reference is preferred at any time when you're sure the shared_ptr itself (not the object it points to) is only accessed from a single thread.

sipa commented on src/main.cpp:
```diff
@@ -5846,7 +5841,8 @@ bool static ProcessMessage(CNode* pfrom, string strCommand, CDataStream& vRecv,
 
-        CBlock block;
+        CBlock *pblock = new CBlock();
+        std::shared_ptr<const CBlock> shared_pblock(pblock); // pblock will be delted by the shared_ptr
```
Typo: deleted

sipa commented on src/main.cpp:
```diff
@@ -2959,14 +2968,16 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
                     state = CValidationState();
                     fInvalidFound = true;
                     fContinue = false;
+                    // If we didn't actually connect the block, don't notify listeners about it
```
No, cs_main is held during the entirety of the attempted reorg here, and the callbacks only fire after the whole thing (I think).

sipa replied:
It seems that failed reorgs are an exception to what I said before, so ignore what I said. I think we should fix this (it could lead to GBT working on the forking point of a failed reorg), but that's outside of the scope here.

sipa commented:
Before headers first, reorgs were dealt with by accumulating the changes in
another cache level, which was only flushed if the reorg succeeded.

Now there is much less a concept of reorgs, and only attempts to improve
the tip. ABC finds what block we want and like to be the tip, and calls
ABCS to make progress towards that. In case of failure, we find what other
block we'd like to become the tip (which may be the previous one), and make
progress towards that. cs_main is only released once we're in a strictly
better state than what we started off in, as otherwise you'd briefly expose
the forking point to GBT during a reorg.

sipa commented on src/main.cpp:
```diff
@@ -2800,28 +2800,40 @@ static int64_t nTimeChainState = 0;
-        if (!ReadBlockFromDisk(block, pindexNew, chainparams.GetConsensus()))
+        std::shared_ptr<CBlock> pblockNew = std::make_shared<CBlock>();
+        connectTrace.blocksConnected.emplace_back(pindexNew, pblockNew);
+        if (!ReadBlockFromDisk(*pblockNew, pindexNew, chainparams.GetConsensus()))
```
Ultranit: if you swap this line with the one above, you can pass std::move(pblockNew) to emplace_back, avoiding an atomic increment + decrement when pblockNew goes out of scope.

TheBlueMatt replied:
No, because pblockNew is a std::atomic<CBlock> not an std::atomic<const CBlock>, so you must call the std::atomic constructor to convert it.

sipa replied:
I see!

JeremyRubin replied:
(if anyone else gets confused, Matt meant std::shared_ptr not std::atomic above)

sipa reviewed (commented):
utACK 3f78bc176a3c612f320bce4516aadcd6234c0850. My nit is just a suggestion.

sipa commented:
Needs rebase.

TheBlueMatt commented:
Rebased.

fanquake commented:
Travis failures:
```
  CXX      libbitcoin_server_a-rest.o
../../src/main.cpp: In function ‘bool ProcessMessage(CNode*, std::string, CDataStream&, int64_t, const CChainParams&, CConnman&)’:
../../src/main.cpp:6094:52: error: ‘block’ was not declared in this scope
             forceProcessing |= MarkBlockAsReceived(block.GetHash());
```
```
  CXX      libbitcoin_server_a-torcontrol.o
../../src/main.cpp: In function ‘bool ProcessMessage(CNode*, std::string, CDataStream&, int64_t, const CChainParams&, CConnman&)’:
../../src/main.cpp:6094:52: error: ‘block’ was not declared in this scope
             forceProcessing |= MarkBlockAsReceived(block.GetHash());
```
```
`main.cpp
../../src/main.cpp: In function ‘bool ProcessMessage(CNode*, std::string, CDataStream&, int64_t, const CChainParams&, CConnman&)’:
../../src/main.cpp:6094:52: error: ‘block’ was not declared in this scope
             forceProcessing |= MarkBlockAsReceived(block.GetHash());
```

paveljanik commented:
Looks like this is not rebased correctly - GH shows 5 conflicting files (and they match my own tests).

TheBlueMatt commented:
It just needs rebased again, closing for now and will rebase on top of #9075 after that gets merged.

TheBlueMatt commented:
Rebased on  #9075.

TheBlueMatt commented:
Rebased after #9125 was merged. Note that #9125 should have fixed most of the performance regression here (and was likely the source of its major reorg speedup, though it should help in more real-world conditions as well), though this should still help some, and is a good step towards #9027 as well as improving callback concurrency in the future.

sipa commented on src/main.cpp:
```diff
@@ -3120,8 +3123,12 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-            GetMainSignals().SyncTransaction(*std::get<0>(connectTrace.txChanged[i]), std::get<1>(connectTrace.txChanged[i]), std::get<2>(connectTrace.txChanged[i]));
+        for (const auto& pair : connectTrace.blocksConnected) {
+            assert(pair.second);
+            const CBlock& block = *(pair.second);
```
Braces are not needed.

sipa commented on src/main.cpp:
```diff
@@ -3110,13 +3118,17 @@ bool ActivateBestChain(CValidationState &state, const CChainParams& chainparams,
-            GetMainSignals().SyncTransaction(*std::get<0>(txChanged[i]), std::get<1>(txChanged[i]), std::get<2>(txChanged[i]));
+        for (const auto& pair : connectTrace.blocksConnected) {
+            assert(pair.second);
+            const CBlock& block = *(pair.second);
```
This loop could be rewritten as 
```
for (const auto& ptx : pair.second->vtx)
    GetMainSignals().SyncTransaction(*ptx, pair.first, i);
```

sipa reviewed (commented):
utACK

gmaxwell commented:
utACK

sdaftuar commented on src/main.cpp:
```diff
@@ -3018,6 +3019,8 @@ static bool ActivateBestChainStep(CValidationState& state, const CChainParams& c
                     fInvalidFound = true;
                     fContinue = false;
+                    // If we didn't actually connect the block, don't notify listeners about it
+                    connectTrace.blocksConnected.pop_back();
```
nit: slightly weird that the caller of ConnectTip() has to know to do this on failures.  Perhaps document this somewhere?

sdaftuar commented:
Code review ACK, will test.

sdaftuar commented:
ACK 7b717cd

TheBlueMatt committed:
Document ConnectBlock connectTrace postconditions

TheBlueMatt committed:
Switch pblock in ProcessNewBlock to a shared_ptr

This (finally) fixes a performance regression in
b3b3c2a5623d5c942d2b3565cc2d833c65105555

TheBlueMatt commented:
Rebased after #9260.
